pkgs/development/interpreters/spidermonkey/common.nix

```
@@ -0,0 +1,184 @@
+{ version, hash }:
+
+{ callPackage
+, lib
+, stdenv
+, fetchurl
+, fetchpatch
+
+# build time
+, buildPackages
+, cargo
+, m4
+, perl
+, pkg-config
+, python3
+, python39
+, rustc
+, which
+, zip
+, autoconf213
+, yasm
+, xcbuild
+
+# runtime
+, icu
+, icu67
+, nspr
+, readline
+, zlib
+, libobjc
+, libiconv
+}:
+
+stdenv.mkDerivation (finalAttrs: rec {
+  pname = "spidermonkey";
+  inherit version;
+
+  src = fetchurl {
+    url = "mirror://mozilla/firefox/releases/${version}esr/source/firefox-${version}esr.source.tar.xz";
+    inherit hash;
+  };
+
+  patches = lib.optionals (lib.versionAtLeast version "102") [
+    # use pkg-config at all systems
+    ./always-check-for-pkg-config.patch
+    ./allow-system-s-nspr-and-icu-on-bootstrapped-sysroot.patch
+  ] ++ lib.optionals (lib.versionOlder version "91") [
+    # Fix build failure on armv7l using Debian patch
+    # Upstream bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1526653
+    (fetchpatch {
+      url = "https://salsa.debian.org/mozilla-team/firefox/commit/fd6847c9416f9eebde636e21d794d25d1be8791d.patch";
+      hash = "sha512-K8U3Qyo7g4si2r/8kJdXyRoTrDHAY48x/YJ7YL+YBwlpfNQcHxX+EZvhRzW8FHYW+f7kOnJu9QykhE8PhSQ9zQ==";
+    })
+
+    # Remove this when updating to 79 - The patches are already applied upstream
+    # https://bugzilla.mozilla.org/show_bug.cgi?id=1318905
+
+    # Combination of 3 changesets, modified to apply on 78:
+    # - https://hg.mozilla.org/mozilla-central/rev/06d7e1b6b7e7
+    # - https://hg.mozilla.org/mozilla-central/rev/ec48f15d085c
+    # - https://hg.mozilla.org/mozilla-central/rev/6803dda74d33
+    ./add-riscv64-support.patch
+  ];
+
+  outputs = [ "out" "dev" ];
+  setOutputFlags = false; # Configure script only understands --includedir
+
+  nativeBuildInputs = [
+    cargo
+    m4
+    perl
+    pkg-config
+    # 78 requires python up to 3.9
+    (if lib.versionOlder version "91" then python39 else python3)
+    rustc
+    rustc.llvmPackages.llvm # for llvm-objdump
+    which
+    zip
+  ] ++ lib.optionals (lib.versionOlder version "91") [
+    autoconf213
+    yasm # to buid icu? seems weird
+  ] ++ lib.optionals stdenv.isDarwin [
+    xcbuild
+    libobjc
+    libiconv
+  ];
+
+  buildInputs = [
+    (if lib.versionOlder version "91" then icu67 else icu)
+    nspr
+    readline
+    zlib
+  ];
+
+  preConfigure = ''
+    ${lib.optionalString (lib.versionOlder version "91") ''
+    export CXXFLAGS="-fpermissive"
+    ''}
+    export LIBXUL_DIST=$out
+    export PYTHON="${buildPackages.python3.interpreter}"
+
+    ${lib.optionalString (lib.versionAtLeast version "91") ''
+    export M4=m4
+    export AWK=awk
+    export AS=$CC
+    export AC_MACRODIR=$PWD/build/autoconf/
+
+    pushd js/src
+    sh ../../build/autoconf/autoconf.sh --localdir=$PWD configure.in > configure
+    chmod +x configure
+    popd
+    ''}
+
+    # We can't build in js/src/, so create a build dir
+    mkdir obj
+    cd obj/
+    configureScript=../js/src/configure
```

```suggestion
    export configureScript=../js/src/configure
```
Shouldn't this be set in nix?